### PR TITLE
[BE] Rewrite test_checkpiont.py

### PR DESCRIPTION
**Pull request by fegin**
The original `test_checkpoint.py` had some issues - it mocked the ConfigManager, which made it prone to breaking.
This PR refactors `test_checkpoint.py` to make it more robust and cover additional logic. While there is still room for improvement, this update addresses some of the existing gaps in test coverage.

### CODE REVIEW

File: tests/unit_tests/test_checkpoint.py [hunk 1]
```diff
@@ -9,341 +9,470 @@
...
+            ft_manager=ft_manager,
         )
-        # First save: should schedule an async save.
+
+        # First save schedules async
         manager.save(curr_step=10, force=False)
-        f = manager.async_future
-        f.result.assert_not_called()
+        future = manager.async_future
+        future.result.assert_not_called()
```
Comment (tianyu-l) [inline tests/unit_tests/test_checkpoint.py:423]: I'm curious about the behavior here.
Why it's guaranteed that the `dcp.async_save` won't be actually called before the next `manager.save` happens? My intuition had been
- before next checkpoint save, we won't know it has been called or not
- if next checkpoint save happens, then the previous save definitely happened

File: tests/unit_tests/test_checkpoint.py [hunk 2]
```diff
@@ -9,341 +9,470 @@
...
 
-def fake_get_model_state_dict(model, *args, **kwargs):
-    return model.state_dict()
+    def load_state_dict(self, sd: dict):
+        pass
 
 
-# TODO: The unittest is not well structured and does not cover enough paths.
-# It should be refactored.
+class FakeDataLoader(DataLoader):
```
Comment (tianyu-l) [inline tests/unit_tests/test_checkpoint.py:51]: just for my understanding: what's the advantage of "fake" real objects over "dummy" mock objects?

File: tests/unit_tests/test_checkpoint.py (hunk 1 above)
Comment (fegin) [reply tests/unit_tests/test_checkpoint.py:423]: `dcp.async_save()` is called when `manager.save()`is called. But we need to guard the NEXT `dcp.async_save()` by checking and waiting the `async_future`. This test verifies this logic.

1. First `async_save` is called and `async_future` is produced but not checked (assert_not_called)
2. Second `async_save` is called and the original `async_future` is checked (assert_called_once)
3. The new `async_future` is produced by the second `async_save` and it is not called.

File: tests/unit_tests/test_checkpoint.py (hunk 2 above)
Comment (fegin) [reply tests/unit_tests/test_checkpoint.py:51]: I cannot say for all the use cases. But I typically just use them together. I wrote a fake class/object that behaves what I expect and use mock to replace them. But sometimes, the mocked behavior is pretty simple (return certain values), then I will probably just use mock.

File: tests/unit_tests/test_checkpoint.py [hunk 3]
```diff
@@ -9,341 +9,470 @@
...
-class DummyJobConfig:
-    checkpoint: DummyCheckpointConfig = field(default_factory=DummyCheckpointConfig)
-    job: DummyJob = field(default_factory=DummyJob)
-    fault_tolerance: DummyFaultTolerance = field(default_factory=DummyFaultTolerance)
-    ft_manager = None
+class DummyFuture:
+    def __init__(self):
+        self.result = mock.Mock()
+
+    def __await__(self):
```
Comment (fegin) [inline tests/unit_tests/test_checkpoint.py]: We can remove this.

**Review by tianyu-l — APPROVED** [review]
LGTM
